File: src/Preferences.m
//
//  Preferences.m
//  Travis CI
//
//  Created by Henrik Hodne on 5/18/12.
//  Copyright (c) 2012 Travis CI GmbH. All rights reserved.
//

#import "Preferences.h"

static NSString * const kRepositoriesSetting = @"repositories";
static NSString * const kFirehoseSetting = @"firehose";
static NSString * const kSelfNotificationsSetting = @"self_notifications";
static NSString * const kLoggedInAsSetting = @"logged_in_as";

@implementation Preferences

+ (Preferences *)sharedPreferences {
  static Preferences *_sharedPreferences = nil;
  static dispatch_once_t onceToken;
  dispatch_once(&onceToken, ^{
    _sharedPreferences = [[Preferences alloc] init];
  });

  return _sharedPreferences;
}

- (void)setupDefaults {
  NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
  [defaults registerDefaults:@{
    kRepositoriesSetting: @[],
    kFirehoseSetting: @(NO),
    kSelfNotificationsSetting: @(YES),
  }];
}

- (NSArray *)repositories {
  NSUserDefaults *userDefault = [NSUserDefaults standardUserDefaults];
  return [userDefault stringArrayForKey:kRepositoriesSetting];
}

- (void)addRepository:(NSString *)slug {
  if (![[self repositories] containsObject:slug]) {
    NSArray *repositories = [[self repositories] arrayByAddingObject:slug];
    [self setRepositories:repositories];
  }
}

- (void)removeRepository:(NSString *)slug {
  NSMutableArray *repositories = [[self repositories] mutableCopy];
  [repositories removeObject:slug];
  [self setRepositories:repositories];
}

- (void)setRepositories:(NSArray *)repositories {
  NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
  [userDefaults setObject:repositories forKey:kRepositoriesSetting];
  [userDefaults synchronize];
}

- (BOOL)firehoseEnabled {
  NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
  return [userDefaults boolForKey:kFirehoseSetting];
}

- (void)setFirehoseEnabled:(BOOL)firehoseEnabled {
  NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
  [userDefaults setBool:firehoseEnabled forKey:kFirehoseSetting];
  [userDefaults synchronize];
}

- (BOOL)selfNotifications {
  NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
  return [userDefaults boolForKey:kSelfNotificationsSetting];
}

- (void)setSelfNotifications:(BOOL)selfNotifications {
  NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
  [userDefaults setBool:selfNotifications forKey:kSelfNotificationsSetting];
  [userDefaults synchronize];
}

- (NSString *)loggedInAs {
  NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
  return [userDefaults objectForKey:kLoggedInAsSetting];
}

- (void)setLoggedInAs:(NSString *)loggedInAs {
  NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
  [userDefaults setObject:loggedInAs forKey:kLoggedInAsSetting];
  [userDefaults synchronize];
}

@end

File: app/models/preferences.rb
class Preferences
  def self.sharedPreferences
    Dispatch.once do
      @@sharedPreferences = Preferences.new(NSUserDefaults.standardUserDefaults)
    end

    @@sharedPreferences
  end

  def initialize(userDefaults)
    @user_defaults = userDefaults
  end

  def setup_defaults
    @user_defaults.registerDefaults({
      "repositories" => [],
      "firehose" => false,
    })
  end

  def repositories
    @user_defaults.stringArrayForKey("repositories") || []
  end

  def add_repository(slug)
    unless self.repositories.containsObject(slug)
      self.repositories = self.repositories.arrayByAddingObject(slug)
    end
  end

  def remove_repository(slug)
    repositories = self.repositories.mutableCopy
    repositories.removeObject(slug)
    self.repositories = repositories
  end

  def firehose_enabled?
    @user_defaults.boolForKey("firehose")
  end

  def firehose_enabled=(enable)
    @user_defaults.setBool(enable, forKey: "firehose")
    @user_defaults.synchronize
  end

  def access_token
    @user_defaults.stringForKey("access_token")
  end

  def access_token=(access_token)
    @user_defaults.setObject(access_token, forKey: "access_token")
    @user_defaults.synchronize
  end

  private

  def repositories=(repositories)
    @user_defaults.setObject(repositories, forKey: "repositories")
    @user_defaults.synchronize
  end
end
